app/src/pages/p/[projectSlug]/usage/index.tsx

```diff
@@ -16,72 +16,240 @@ import {
   Tr,
   VStack,
   Box,
+  Spacer,
+  Tabs,
+  TabList,
+  Tab,
+  TabPanels,
+  TabPanel,
 } from "@chakra-ui/react";
-import { DollarSign, Hash } from "lucide-react";
+import { ChevronRightIcon, ChevronLeftIcon, DollarSign, Hash } from "lucide-react";
+import { useRouter } from "next/router";
+import { useEffect, useMemo } from "react";
+import { useQueryParam, StringParam } from "use-query-params";
 import { ProjectLink } from "~/components/ProjectLink";
+import CostGraph from "~/components/dashboard/CostGraph";
 
 import UsageGraph from "~/components/dashboard/UsageGraph";
 import AppShell from "~/components/nav/AppShell";
 import { modelInfo } from "~/server/fineTuningProviders/supportedModels";
-import { api } from "~/utils/api";
-import dayjs from "~/utils/dayjs";
-import { useSelectedProject } from "~/utils/hooks";
+import dayjs, { formatToUTCDayMonth } from "~/utils/dayjs";
+import { useSelectedProject, useStats } from "~/utils/hooks";
 
 const numberWithDefault = (num: number | string | bigint | null, defaultValue = 0) =>
   Number(num ?? defaultValue);
 
 export default function Usage() {
   const { data: selectedProject } = useSelectedProject();
 
-  const stats = api.usage.stats.useQuery(
-    { projectId: selectedProject?.id ?? "" },
-    { enabled: !!selectedProject },
-  );
+  const router = useRouter();
+
+  const [startDate, setStartDate] = useQueryParam("start", StringParam);
+  const [endDate, setEndDate] = useQueryParam("end", StringParam);
+
+  const startOfThisMonth = dayjs().startOf("month").format("YYYY-MM-DD");
+  const endOfThisMonth = dayjs().endOf("month").format("YYYY-MM-DD");
+
+  const { projectSlug } = router.query;
+
+  useEffect(() => {
+    // Prevents `href` Interpolation error in nextjs
+    if (projectSlug) {
+      if (!startDate) setStartDate(startOfThisMonth);
+      if (!endDate) setEndDate(endOfThisMonth);
+    }
+  }, [startDate, endDate, setStartDate, setEndDate, projectSlug]);
+
+  const stats = useStats(selectedProject?.id || "", startDate, endDate);
+
+  const { totalInferenceSpend, totalTrainingSpend, totalInputTokens, totalOutputTokens } =
+    useMemo(() => {
+      const calculateTotal = <T extends Record<string, any>>(
+        dataArray: T[] | undefined,
+        key: keyof T,
+      ) => dataArray?.reduce((acc: number, cur: T) => acc + Number(cur[key]), 0) ?? 0;
+
+      return {
+        totalTrainingSpend: calculateTotal(stats.data?.periods, "trainingCost"),
```

Comment:
I would calculate these totals on the server and pass them to the client. Then you could get rid of this useMemo. In general, it's best to do computations on the server where possible.

Reply:
Awesome idea! Implemented server-side calculations.